File: Decode Ways.cpp
class Solution {
    public:
        int solve(int i, string s, int n, vector<int>&dp){
            if(i == n)
                return  1;
            if(s[i] == '0' ) 
                return 0;
            if (dp[i] != -1)
                return dp[i];
    
            int ways =  solve(i+1, s, n, dp);
            if(i+1 < n && ( s[i] == '1' || (s[i]=='2' && s[i+1] <='6'))){
                ways += solve(i+2, s, n, dp);
            }
            return dp[i]=ways;
            
        }
        int numDecodings(string s) {
            int n = s.length();
            vector<int> dp(n, -1);
            
            return solve(0, s,n, dp);
        }
    };

File: neetcode/91-decode-ways/decode-ways.java
class Solution {
    public int numDecodings(String s) {
        
        int n = s.length();

        //stores no.of ways to decode n length string
        int [] dp = new int [n + 1];

        dp[0] = 1; //decode as invalid;

        for(int i = 1; i <= n; ++i){
            
            //considering as single element
            //i : s[0...i-1] ==> 
            if(s.charAt(i-1) != '0') dp[i] = dp[i-1];

            //decoding as 2 digits ==> no trailing 0's && <= 26
            //length should be atleast 2
            //previous element must not be = 0
            // i: s[0....i-1]i
            if(i > 1 && s.charAt(i-2) != '0'){
                //substring(i,j) =>s[i , j-1]
                //
                int twoDigitNumber = Integer.valueOf(s.substring(i-2, i));
                if(twoDigitNumber <= 26) dp[i] += dp[i-2];
            }
        }
        return dp[n];

    }
}